### PR TITLE
Acquisition aggregation improvements

### PR DESCRIPTION
**Pull request by javdrher**
Avoids tree structures with great depth with summing or multiplying many acquisition functions. This also reduces the size of the TF graph (by using reduce operations rather than many additions/multiplications). Useful for later implementation of #18 .

### CODE REVIEW

File: GPflowOpt/acquisition.py [hunk 1]
```diff
@@ -168,47 +170,63 @@ def build_acquisition(self, Xcand):
...
-        super(AcquisitionBinaryOperator, self).__init__()
-        assert isinstance(lhs, Acquisition)
-        assert isinstance(rhs, Acquisition)
-        self.lhs = lhs
-        self.rhs = rhs
+class AcquisitionAggregationOperator(Acquisition):
+    def __init__(self, operands, oper):
+        super(AcquisitionAggregationOperator, self).__init__()
+        for operand in operands:
+            assert isinstance(operand, Acquisition)
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: a shorter way:
assert( all(isinstance(x, int) for x in lst) )

?

File: GPflowOpt/acquisition.py [hunk 2]
```diff
@@ -168,47 +170,63 @@ def build_acquisition(self, Xcand):
...
+        self.operands = ParamList(operands)
         self._oper = oper
 
     @Acquisition.data.getter
     def data(self):
-        lhsX, lhsY = self.lhs.data
-        rhsX, rhsY = self.rhs.data
+        X = self.operands[0].data[0]
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: shouldnt we check if X is the same for all operands?

File: GPflowOpt/acquisition.py [hunk 3]
```diff
@@ -168,47 +170,63 @@ def build_acquisition(self, Xcand):
...
-            return lhsX, np.hstack((lhsY, rhsY))
+            return X, np.hstack(Ys)
 
     def set_data(self, X, Y):
-        offset_lhs = self.lhs.set_data(X, Y)
-        offset_rhs = self.rhs.set_data(X, Y[:, offset_lhs:])
-        return offset_lhs + offset_rhs
+        offset = 0
+        for operand in self.operands:
+            offset += operand.set_data(X, Y[:, offset:])
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: I'm guessing this needs to be changed for multi-objective criteria? Taking in multiple columns of Y

File: GPflowOpt/acquisition.py [hunk 4]
```diff
@@ -168,47 +170,63 @@ def build_acquisition(self, Xcand):
...
-    Base class for implementing binary operators for acquisition functions, for defining joint acquisition functions
-    """
-
-    def __init__(self, lhs, rhs, oper):
-        super(AcquisitionBinaryOperator, self).__init__()
-        assert isinstance(lhs, Acquisition)
-        assert isinstance(rhs, Acquisition)
-        self.lhs = lhs
-        self.rhs = rhs
+class AcquisitionAggregationOperator(Acquisition):
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: Actually this is quite a long name. anything shorter? although we are living the world of autocompletion, something like AcquisitionAggregator might be slightly easier to remember?

Comment (icouckuy) [inline GPflowOpt/acquisition.py]: then again, you would only use the derived classes ;)

File: testing/test_acquisition.py [hunk 5]
```diff
@@ -145,33 +149,33 @@ def test_PoF_validity(self):
         self.assertTrue(np.allclose(self.acquisition.evaluate(X2), 0), msg="Right half of plane is not feasible")
 
 
-class _TestAcquisitionBinaryOperator(_TestAcquisition):
+class _TestAcquisitionAggregationOperator(_TestAcquisition):
```
Comment (icouckuy) [inline testing/test_acquisition.py]: drop the Operator

File: GPflowOpt/acquisition.py (hunk 3 above)
Comment (javdrher) [inline GPflowOpt/acquisition.py]: set_data implementations can return offset > 1, so multiple columns are skipped for the next operand

File: GPflowOpt/acquisition.py [hunk 6]
```diff
@@ -112,10 +112,14 @@ def evaluate(self, Xcand):
...
     def __add__(self, other):
-        return AcquisitionSum(self, other)
+        if isinstance(other, AcquisitionSum):
+            return AcquisitionSum([self] + other.operands.sorted_params)
+        return AcquisitionSum([self, other])
 
     def __mul__(self, other):
-        return AcquisitionProduct(self, other)
+        if isinstance(other, AcquisitionProduct):
+            return AcquisitionProduct([self] + other.operands.sorted_params)
```
Comment (javdrher) [inline GPflowOpt/acquisition.py]: This was just most convenient to preserve the order

File: GPflowOpt/acquisition.py (hunk 2 above)
Comment (javdrher) [inline GPflowOpt/acquisition.py]: k, adding this. However, only for non-tf mode.

**Pull request merged**